File: src/v2/codec.ts
import { FiseError } from "../errors.js";
import { assertBytes, copyBytes, isBytes, snapshotBytes } from "./bytes.js";
import { compressLz4Block, decompressLz4Block } from "./lz4.js";
import type {
	FiseContext,
	FiseContextValue,
	FiseJsonValue,
	FiseValue
} from "./types.js";

const METADATA_VERSION = 1;
const STRUCTURED_DATA = 1;
const BINARY_DATA = 2;
const COMPRESSED_STRUCTURED_DATA = 3;
export const FISE_VALUE_METADATA_LENGTH = 2;
const COMPRESSED_STRUCTURED_HEADER_LENGTH = FISE_VALUE_METADATA_LENGTH + 4;
const MINIMUM_COMPRESSION_INPUT = 256;
const MAX_STRUCTURED_CONTENT_LENGTH = 512 * 1024 * 1024;
const MAX_COMPRESSION_RATIO = 256;
const MAX_NESTING_DEPTH = 64;
const MAX_CONTEXT_BYTES = 64 * 1024;
const BASE64_URL_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

const encoder = new TextEncoder();

export function encodeValue(value: unknown): Uint8Array {
	if (isBytes(value)) {
		const bytes = snapshotBytes(value, "binary input", "INVALID_INPUT");
		const output = new Uint8Array(FISE_VALUE_METADATA_LENGTH + bytes.length);
		output[0] = METADATA_VERSION;
		output[1] = BINARY_DATA;
		output.set(bytes, FISE_VALUE_METADATA_LENGTH);
		return output;
	}

	const canonical = canonicalJson(value, "input");
	const content = encoder.encode(canonical);
	if (content.length >= MINIMUM_COMPRESSION_INPUT && content.length <= 0xffff_ffff) {
		const compressed = compressLz4Block(content);
		if (compressed.length + 4 < content.length) {
			const output = new Uint8Array(
				COMPRESSED_STRUCTURED_HEADER_LENGTH + compressed.length
			);
			output[0] = METADATA_VERSION;
			output[1] = COMPRESSED_STRUCTURED_DATA;
			new DataView(output.buffer).setUint32(
				FISE_VALUE_METADATA_LENGTH,
				content.length,
				false
			);
			output.set(compressed, COMPRESSED_STRUCTURED_HEADER_LENGTH);
			return output;
		}
	}
	const output = new Uint8Array(FISE_VALUE_METADATA_LENGTH + content.length);
	output[0] = METADATA_VERSION;
	output[1] = STRUCTURED_DATA;
	output.set(content, FISE_VALUE_METADATA_LENGTH);
	return output;
}

export function decodeValue(payload: Uint8Array): FiseValue {
	assertBytes(payload, "restored payload", "INVALID_PAYLOAD");
	assertPayloadMetadata(payload);
	if (payload[1] === BINARY_DATA) {
		return copyBytes(payload.subarray(FISE_VALUE_METADATA_LENGTH));
	}
	let content: Uint8Array;
	if (payload[1] === STRUCTURED_DATA) {
		content = payload.subarray(FISE_VALUE_METADATA_LENGTH);
	} else if (payload[1] === COMPRESSED_STRUCTURED_DATA) {
		content = decodeCompressedStructured(payload);
	} else {
		throw new FiseError("INVALID_PAYLOAD", `FISE: unknown data type ${payload[1]}.`);
	}

	let source: string;
	try {
		source = new TextDecoder("utf-8", { fatal: true }).decode(content);
	} catch (error) {
		throw new FiseError("INVALID_PAYLOAD", "FISE: structured payload is not valid UTF-8.", error);
	}
	let value: unknown;
	try {
		value = JSON.parse(source);
	} catch (error) {
		throw new FiseError("INVALID_PAYLOAD", "FISE: structured payload is not valid JSON.", error);
	}
	let canonical: string;
	try {
		canonical = canonicalJson(value, "restored payload");
	} catch (error) {
		throw new FiseError("INVALID_PAYLOAD", "FISE: restored JSON is outside the data contract.", error);
	}
	if (source !== canonical) {
		throw new FiseError("INVALID_PAYLOAD", "FISE: structured payload is not canonical JSON.");
	}
	return value as FiseJsonValue;
}

function decodeCompressedStructured(payload: Uint8Array): Uint8Array {
	if (payload.length < COMPRESSED_STRUCTURED_HEADER_LENGTH + 1) {
		throw new FiseError("INVALID_PAYLOAD", "FISE: compressed structured payload is truncated.");
	}
	const originalLength = new DataView(
		payload.buffer,
		payload.byteOffset,
		payload.byteLength
	).getUint32(FISE_VALUE_METADATA_LENGTH, false);
	if (originalLength > MAX_STRUCTURED_CONTENT_LENGTH) {
		throw new FiseError(
			"INVALID_PAYLOAD",
			`FISE: restored structured payload exceeds ${MAX_STRUCTURED_CONTENT_LENGTH} bytes.`
		);
	}
	const compressedLength = payload.length - COMPRESSED_STRUCTURED_HEADER_LENGTH;
	if (
		originalLength > Math.max(
			MINIMUM_COMPRESSION_INPUT,
			compressedLength * MAX_COMPRESSION_RATIO
		)
	) {
		throw new FiseError(
			"INVALID_PAYLOAD",
			"FISE: compressed structured payload exceeds the allowed expansion ratio."
		);
	}
	return decompressLz4Block(
		payload.subarray(COMPRESSED_STRUCTURED_HEADER_LENGTH),
		originalLength
	);
}

/** Validates the transformed prefix required by direct binary restoration. */
export function assertBinaryPayloadMetadata(payloadPrefix: Uint8Array): void {
	assertBytes(payloadPrefix, "restored payload metadata", "INVALID_PAYLOAD");
	assertPayloadMetadata(payloadPrefix);
	if (payloadPrefix[1] !== BINARY_DATA) {
		throw new FiseError(
			"INVALID_PAYLOAD",
			"FISE: range and progressive restoration require a top-level binary envelope."
		);
	}
}

function assertPayloadMetadata(payload: Uint8Array): void {
	if (payload.length < FISE_VALUE_METADATA_LENGTH) {
		throw new FiseError("INVALID_PAYLOAD", "FISE: restored payload is missing metadata.");
	}
	if (payload[0] !== METADATA_VERSION) {
		throw new FiseError(
			"INVALID_PAYLOAD",
			`FISE: unsupported payload metadata version ${payload[0]}.`
		);
	}
}

export interface PreparedContext {
	readonly value: FiseContext;
	readonly encoded: Uint8Array;
}

export function prepareContext(context: unknown): PreparedContext {
	const source = canonicalJson(context === undefined ? [] : context, "context");
	let parsed: unknown;
	try {
		parsed = JSON.parse(source);
	} catch (error) {
		throw new FiseError("INVALID_CONTEXT", "FISE: unable to snapshot context.", error);
	}
	if (!Array.isArray(parsed)) {
		throw invalidValue("context", "must be a positional array");
	}
	const values: FiseContextValue[] = [];
	for (let index = 0; index < parsed.length; index++) {
		const value = parsed[index];
		if (
			value !== null &&
			typeof value !== "boolean" &&
			typeof value !== "number" &&
			typeof value !== "string"
		) {
			throw invalidValue(`context[${index}]`, "must be a JSON scalar");
		}
		values.push(value as FiseContextValue);
	}
	const canonicalBytes = encoder.encode(source);
	if (canonicalBytes.length > MAX_CONTEXT_BYTES) {
		throw new FiseError(
			"INVALID_CONTEXT",
			`FISE: canonical context exceeds ${MAX_CONTEXT_BYTES} bytes.`
		);
	}
	return Object.freeze({
		value: Object.freeze(values),
		encoded: encodeBase64Url(canonicalBytes)
	});
}

function encodeBase64Url(input: Uint8Array): Uint8Array {
	const outputLength = Math.floor((input.length * 4 + 2) / 3);
	const output = new Uint8Array(outputLength);
	let inputOffset = 0;
	let outputOffset = 0;
	while (inputOffset + 2 < input.length) {
		const value =
			(input[inputOffset] << 16) |
			(input[inputOffset + 1] << 8) |
			input[inputOffset + 2];
		output[outputOffset++] = BASE64_URL_ALPHABET.charCodeAt((value >>> 18) & 63);
		output[outputOffset++] = BASE64_URL_ALPHABET.charCodeAt((value >>> 12) & 63);
		output[outputOffset++] = BASE64_URL_ALPHABET.charCodeAt((value >>> 6) & 63);
		output[outputOffset++] = BASE64_URL_ALPHABET.charCodeAt(value & 63);
		inputOffset += 3;
	}
	const remaining = input.length - inputOffset;
	if (remaining === 1) {
		const value = input[inputOffset] << 16;
		output[outputOffset++] = BASE64_URL_ALPHABET.charCodeAt((value >>> 18) & 63);
		output[outputOffset] = BASE64_URL_ALPHABET.charCodeAt((value >>> 12) & 63);
	} else if (remaining === 2) {
		const value = (input[inputOffset] << 16) | (input[inputOffset + 1] << 8);
		output[outputOffset++] = BASE64_URL_ALPHABET.charCodeAt((value >>> 18) & 63);
		output[outputOffset++] = BASE64_URL_ALPHABET.charCodeAt((value >>> 12) & 63);
		output[outputOffset] = BASE64_URL_ALPHABET.charCodeAt((value >>> 6) & 63);
	}
	return output;
}

export function canonicalJson(value: unknown, label = "value"): string {
	const ancestors = new Set<object>();
	try {
		const canonical = encodeCanonical(value, label, 0, ancestors);
		assertNativePlainGraph(value, label);
		return canonical;
	} catch (error) {
		if (error instanceof FiseError) throw error;
		throw new FiseError(
			invalidCode(label),
			`FISE: unable to canonicalize ${label}.`,
			error
		);
	}
}

function encodeCanonical(
	value: unknown,
	label: string,
	depth: number,
	ancestors: Set<object>
): string {
	if (depth > MAX_NESTING_DEPTH) {
		throw invalidValue(label, `nesting exceeds ${MAX_NESTING_DEPTH}`);
	}
	if (value === null) return "null";
	if (typeof value === "string") {
		assertUnicodeScalarString(value, label);
		return JSON.stringify(value);
	}
	if (typeof value === "boolean") return JSON.stringify(value);
	if (typeof value === "number") {
		if (!Number.isFinite(value) || Object.is(value, -0)) {
			throw invalidValue(label, "numbers must be finite and must not be negative zero");
		}
		return JSON.stringify(value);
	}
	if (typeof value !== "object") {
		throw invalidValue(label, "must contain only JSON-safe values");
	}
	if (isBytes(value)) {
		throw invalidValue(label, "may contain Uint8Array only as the top-level input");
	}
	if (ancestors.has(value)) throw invalidValue(label, "must not contain cycles");
	ancestors.add(value);
	try {
		if (Array.isArray(value)) {
			return encodeArray(value, label, depth, ancestors);
		}
		return encodeObject(value, label, depth, ancestors);
	} finally {
		ancestors.delete(value);
	}
}

function encodeArray(
	value: readonly unknown[],
	label: string,
	depth: number,
	ancestors: Set<object>
): string {
	if (!isSupportedArrayPrototype(Object.getPrototypeOf(value))) {
		throw invalidValue(label, "arrays must have Array.prototype");
	}
	const keys = Reflect.ownKeys(value);
	for (const key of keys) {
		if (typeof key === "symbol") throw invalidValue(label, "must not contain symbol keys");
		if (key === "length") continue;
		if (!/^(?:0|[1-9]\d*)$/.test(key) || Number(key) >= value.length) {
			throw invalidValue(label, "arrays must not contain custom properties");
		}
	}
	const parts: string[] = [];
	for (let index = 0; index < value.length; index++) {
		const descriptor = Object.getOwnPropertyDescriptor(value, String(index));
		if (!descriptor || !("value" in descriptor)) {
			throw invalidValue(label, "arrays must be dense and must not contain accessors");
		}
		if (!descriptor.enumerable) {
			throw invalidValue(label, "array elements must be enumerable");
		}
		parts.push(encodeCanonical(descriptor.value, `${label}[${index}]`, depth + 1, ancestors));
	}
	return `[${parts.join(",")}]`;
}

function encodeObject(
	value: object,
	label: string,
	depth: number,
	ancestors: Set<object>
): string {
	const prototype = Object.getPrototypeOf(value);
	if (!isSupportedObjectPrototype(prototype)) {
		throw invalidValue(label, "objects must have Object.prototype or null prototype");
	}
	const ownKeys = Reflect.ownKeys(value);
	if (ownKeys.some(key => typeof key === "symbol")) {
		throw invalidValue(label, "must not contain symbol keys");
	}
	const keys = ownKeys as string[];
	for (const key of keys) assertUnicodeScalarString(key, `${label} property name`);
	keys.sort();
	const parts: string[] = [];
	for (const key of keys) {
		const descriptor = Object.getOwnPropertyDescriptor(value, key);
		if (!descriptor || !("value" in descriptor)) {
			throw invalidValue(label, "must not contain accessors");
		}
		if (!descriptor.enumerable) {
			throw invalidValue(label, "must contain only enumerable properties");
		}
		parts.push(
			`${JSON.stringify(key)}:${encodeCanonical(descriptor.value, `${label}.${key}`, depth + 1, ancestors)}`
		);
	}
	return `{${parts.join(",")}}`;
}

function assertUnicodeScalarString(value: string, label: string): void {
	for (let index = 0; index < value.length; index++) {
		const codeUnit = value.charCodeAt(index);
		if (codeUnit >= 0xd800 && codeUnit <= 0xdbff) {
			const next = value.charCodeAt(index + 1);
			if (!(next >= 0xdc00 && next <= 0xdfff)) {
				throw invalidValue(label, "must contain only valid Unicode scalar values");
			}
			index++;
			continue;
		}
		if (codeUnit >= 0xdc00 && codeUnit <= 0xdfff) {
			throw invalidValue(label, "must contain only valid Unicode scalar values");
		}
	}
}

function isSupportedObjectPrototype(prototype: object | null): boolean {
	if (prototype === null || prototype === Object.prototype) return true;
	if (Object.getPrototypeOf(prototype) !== null) return false;

	const constructorDescriptor = Object.getOwnPropertyDescriptor(prototype, "constructor");
	if (
		!constructorDescriptor ||
		!("value" in constructorDescriptor) ||
		constructorDescriptor.enumerable ||
		typeof constructorDescriptor.value !== "function"
	) {
		return false;
	}
	const constructor = constructorDescriptor.value;
	if (constructor.name !== "Object" || constructor.prototype !== prototype) return false;
	try {
		const source = Function.prototype.toString.call(constructor);
		if (!/^function Object\(\)\s*\{\s*\[native code\]\s*\}$/.test(source)) return false;
	} catch {
		return false;
	}
	for (const key of Reflect.ownKeys(prototype)) {
		if (Object.getOwnPropertyDescriptor(prototype, key)?.enumerable !== false) return false;
	}
	return true;
}

function isSupportedArrayPrototype(prototype: object | null): boolean {
	if (prototype === Array.prototype) return true;
	if (prototype === null || !isSupportedObjectPrototype(Object.getPrototypeOf(prototype))) {
		return false;
	}
	const constructorDescriptor = Object.getOwnPropertyDescriptor(prototype, "constructor");
	if (
		!constructorDescriptor ||
		!("value" in constructorDescriptor) ||
		constructorDescriptor.enumerable ||
		typeof constructorDescriptor.value !== "function"
	) {
		return false;
	}
	const constructor = constructorDescriptor.value;
	if (constructor.name !== "Array" || constructor.prototype !== prototype) return false;
	try {
		const source = Function.prototype.toString.call(constructor);
		if (!/^function Array\(\)\s*\{\s*\[native code\]\s*\}$/.test(source)) return false;
	} catch {
		return false;
	}
	for (const key of Reflect.ownKeys(prototype)) {
		if (Object.getOwnPropertyDescriptor(prototype, key)?.enumerable !== false) return false;
	}
	return true;
}

function assertNativePlainGraph(value: unknown, label: string): void {
	if (value === null || typeof value !== "object") return;
	try {
		structuredClone(value);
	} catch (error) {
		throw new FiseError(
			label === "context" ? "INVALID_CONTEXT" : "INVALID_INPUT",
			`FISE: ${label} must not contain Proxy or non-cloneable wrapper objects.`,
			error
		);
	}
}

function invalidValue(label: string, detail: string): FiseError {
	return new FiseError(invalidCode(label), `FISE: ${label} ${detail}.`);
}

function invalidCode(label: string): "INVALID_CONTEXT" | "INVALID_INPUT" {
	return label === "context" ||
		label.startsWith("context.") ||
		label.startsWith("context[") ||
		label.startsWith("context ")
		? "INVALID_CONTEXT"
		: "INVALID_INPUT";
}

File: python/src/fise/_codec.py
from __future__ import annotations

import base64
import json
import math
from decimal import Decimal
from typing import Any

from ._lz4 import compress_lz4_block, decompress_lz4_block
from .errors import FiseError
from .profile_runtime import Context, ContextScalar

METADATA_LENGTH = 2
_METADATA_VERSION = 1
_STRUCTURED_DATA = 1
_BINARY_DATA = 2
_COMPRESSED_STRUCTURED_DATA = 3
_COMPRESSED_HEADER_LENGTH = METADATA_LENGTH + 4
_MINIMUM_COMPRESSION_INPUT = 256
_MAX_STRUCTURED_CONTENT_LENGTH = 512 * 1024 * 1024
_MAX_COMPRESSION_RATIO = 256
_MAX_NESTING_DEPTH = 64
_MAX_CONTEXT_BYTES = 64 * 1024


def encode_value(value: Any) -> bytes:
    if type(value) is bytes:
        return bytes((_METADATA_VERSION, _BINARY_DATA)) + value
    canonical = canonical_json(value, "input")
    content = canonical.encode("utf-8")
    if len(content) >= _MINIMUM_COMPRESSION_INPUT and len(content) <= 0xFFFFFFFF:
        compressed = compress_lz4_block(content)
        if len(compressed) + 4 < len(content):
            return (
                bytes((_METADATA_VERSION, _COMPRESSED_STRUCTURED_DATA))
                + len(content).to_bytes(4, "big")
                + compressed
            )
    return bytes((_METADATA_VERSION, _STRUCTURED_DATA)) + content


def decode_value(payload: bytes) -> Any:
    _assert_payload_metadata(payload)
    data_type = payload[1]
    if data_type == _BINARY_DATA:
        return bytes(payload[METADATA_LENGTH:])
    if data_type == _STRUCTURED_DATA:
        content = payload[METADATA_LENGTH:]
    elif data_type == _COMPRESSED_STRUCTURED_DATA:
        content = _decode_compressed_structured(payload)
    else:
        raise FiseError("INVALID_PAYLOAD", f"FISE: unknown data type {data_type}.")
    try:
        source = content.decode("utf-8", "strict")
    except UnicodeDecodeError as error:
        raise FiseError("INVALID_PAYLOAD", "FISE: structured payload is not valid UTF-8.", error) from error
    try:
        value = json.loads(
            source,
            parse_int=_parse_json_integer,
            parse_constant=_reject_json_constant,
        )
    except (ValueError, TypeError, json.JSONDecodeError) as error:
        raise FiseError("INVALID_PAYLOAD", "FISE: structured payload is not valid JSON.", error) from error
    try:
        canonical = canonical_json(value, "restored payload")
    except FiseError as error:
        raise FiseError("INVALID_PAYLOAD", "FISE: restored JSON is outside the data contract.", error) from error
    if source != canonical:
        raise FiseError("INVALID_PAYLOAD", "FISE: structured payload is not canonical JSON.")
    return value


def assert_binary_payload_metadata(payload_prefix: bytes) -> None:
    _assert_payload_metadata(payload_prefix)
    if payload_prefix[1] != _BINARY_DATA:
        raise FiseError(
            "INVALID_PAYLOAD",
            "FISE: range and progressive restoration require a top-level binary envelope.",
        )


def prepare_context(context: object = None, *, omitted: bool = False) -> tuple[Context, bytes]:
    source_value: object = [] if omitted else context
    source = canonical_json(source_value, "context")
    if type(source_value) is not list:
        raise FiseError("INVALID_CONTEXT", "FISE: context must be a positional list.")
    values: list[ContextScalar] = []
    for index, value in enumerate(source_value):
        if value is None or type(value) in (bool, int, float, str):
            values.append(value)
        else:
            raise FiseError(
                "INVALID_CONTEXT",
                f"FISE: context[{index}] must be a JSON scalar.",
            )
    encoded_source = source.encode("utf-8")
    if len(encoded_source) > _MAX_CONTEXT_BYTES:
        raise FiseError(
            "INVALID_CONTEXT",
            f"FISE: canonical context exceeds {_MAX_CONTEXT_BYTES} bytes.",
        )
    encoded = base64.urlsafe_b64encode(encoded_source).rstrip(b"=")
    return tuple(values), encoded


def canonical_json(value: object, label: str = "value") -> str:
    ancestors: set[int] = set()
    return _encode_canonical(value, label, 0, ancestors)


def _encode_canonical(
    value: object,
    label: str,
    depth: int,
    ancestors: set[int],
) -> str:
    if depth > _MAX_NESTING_DEPTH:
        raise _invalid_value(label, f"nesting exceeds {_MAX_NESTING_DEPTH}")
    if value is None:
        return "null"
    if type(value) is str:
        _assert_unicode_scalar(value, label)
        return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    if type(value) is bool:
        return "true" if value else "false"
    if type(value) in (int, float):
        return _serialize_number(value, label)
    if type(value) is bytes:
        raise _invalid_value(label, "may contain bytes only as the top-level input")
    if type(value) not in (list, dict):
        raise _invalid_value(label, "must contain only JSON-safe values")
    identity = id(value)
    if identity in ancestors:
        raise _invalid_value(label, "must not contain cycles")
    ancestors.add(identity)
    try:
        if type(value) is list:
            return "[" + ",".join(
                _encode_canonical(item, f"{label}[{index}]", depth + 1, ancestors)
                for index, item in enumerate(value)
            ) + "]"
        entries: list[str] = []
        for key in sorted(value, key=_utf16_sort_key):
            if type(key) is not str:
                raise _invalid_value(label, "object keys must be strings")
            _assert_unicode_scalar(key, f"{label} property name")
            encoded_key = json.dumps(key, ensure_ascii=False, separators=(",", ":"))
            entries.append(
                encoded_key
                + ":"
                + _encode_canonical(value[key], f"{label}.{key}", depth + 1, ancestors)
            )
        return "{" + ",".join(entries) + "}"
    finally:
        ancestors.remove(identity)


def _serialize_number(value: int | float, label: str) -> str:
    if type(value) is int:
        try:
            number = float(value)
        except OverflowError as error:
            raise _invalid_value(label, "integers must be exactly representable as binary64", error) from error
        if not math.isfinite(number) or int(number) != value:
            raise _invalid_value(label, "integers must be exactly representable as binary64")
    else:
        number = value
    if not math.isfinite(number) or (number == 0.0 and math.copysign(1.0, number) < 0):
        raise _invalid_value(label, "numbers must be finite and must not be negative zero")
    if number == 0:
        return "0"
    source = repr(number).lower()
    absolute = abs(number)
    if 1e-6 <= absolute < 1e21:
        fixed = format(Decimal(source), "f")
        if "." in fixed:
            fixed = fixed.rstrip("0").rstrip(".")
        return fixed
    if "e" not in source:
        source = format(number, ".17e")
    coefficient, exponent_source = source.split("e", 1)
    if coefficient.endswith(".0"):
        coefficient = coefficient[:-2]
    exponent = int(exponent_source)
    sign = "+" if exponent >= 0 else ""
    return f"{coefficient}e{sign}{exponent}"


def _utf16_sort_key(value: object) -> bytes:
    if type(value) is not str:
        raise _invalid_value("object", "object keys must be strings")
    _assert_unicode_scalar(value, "object property name")
    return value.encode("utf-16-be")


def _assert_unicode_scalar(value: str, label: str) -> None:
    for character in value:
        code_point = ord(character)
        if 0xD800 <= code_point <= 0xDFFF:
            raise _invalid_value(label, "strings must not contain unpaired UTF-16 surrogates")


def _assert_payload_metadata(payload: bytes) -> None:
    if type(payload) is not bytes or len(payload) < METADATA_LENGTH:
        raise FiseError("INVALID_PAYLOAD", "FISE: restored payload is missing metadata.")
    if payload[0] != _METADATA_VERSION:
        raise FiseError(
            "INVALID_PAYLOAD",
            f"FISE: unsupported payload metadata version {payload[0]}.",
        )


def _decode_compressed_structured(payload: bytes) -> bytes:
    if len(payload) < _COMPRESSED_HEADER_LENGTH + 1:
        raise FiseError("INVALID_PAYLOAD", "FISE: compressed structured payload is truncated.")
    original_length = int.from_bytes(payload[METADATA_LENGTH:_COMPRESSED_HEADER_LENGTH], "big")
    if original_length > _MAX_STRUCTURED_CONTENT_LENGTH:
        raise FiseError(
            "INVALID_PAYLOAD",
            f"FISE: restored structured payload exceeds {_MAX_STRUCTURED_CONTENT_LENGTH} bytes.",
        )
    compressed = payload[_COMPRESSED_HEADER_LENGTH:]
    if original_length > max(_MINIMUM_COMPRESSION_INPUT, len(compressed) * _MAX_COMPRESSION_RATIO):
        raise FiseError(
            "INVALID_PAYLOAD",
            "FISE: compressed structured payload exceeds the allowed expansion ratio.",
        )
    return decompress_lz4_block(compressed, original_length)


def _reject_json_constant(value: str) -> None:
    raise ValueError(f"invalid JSON constant {value}")


def _parse_json_integer(source: str) -> int | float:
    integer = int(source)
    try:
        number = float(source)
    except (OverflowError, ValueError):
        return integer
    if (
        math.isfinite(number)
        and int(number) != integer
        and _serialize_number(number, "JSON number") == source
    ):
        return number
    return integer


def _invalid_value(
    label: str,
    reason: str,
    cause: BaseException | None = None,
) -> FiseError:
    code = "INVALID_CONTEXT" if label.startswith("context") else "INVALID_INPUT"
    return FiseError(code, f"FISE: {label} {reason}.", cause)
